### DISCUSSION TITLE
The app works locally but fails on deployment even tho it is built successfully

### DISCUSSION THREAD

**Discussion by aly8i** [vercel/community#958] [Help]
I think I have a lot of questions to ask.

Problem:

My application works fine in local but crashes after deployment

Configuration:

- I am using Heroku to deploy an express auth server.
- I set the environment variables in next js.
- I changed the localhost:3000/api to myapp.vercel.app/api when setting the base URL to the domain given by vercel environment variables.

Output:

- The build works successfully even tho it showed a bunch of warnings.
- the output of my deployed code showed serverless and edge functions that I can't preview with errors.
- I think every thing server generated is failing.
- some of the pages are displaying internal server errors.
![Screenshot 2022-10-16 152120](https://user-images.githubusercontent.com/98344117/196035360-00d9afec-8245-4856-a85b-11be5b1a25fc.jpg)
![Screenshot 2022-10-16 152208](https://user-images.githubusercontent.com/98344117/196035363-b4e97db1-0acc-471a-b501-f7af313630ee.jpg)
![Screenshot 2022-10-16 152255](https://user-images.githubusercontent.com/98344117/196035367-6d4a3d1c-d76e-4efa-89e5-15a67862f6e9.jpg)
![Screenshot 2022-10-16 152527](https://user-images.githubusercontent.com/98344117/196035369-109210fc-c4f3-44f8-a829-b404a852d63d.jpg)

Questions:

- How can I solve my issue?
- Can I deploy the express server within vercel and under the same domain? for example, under different port or /, and how can I achieve that? It will help me force the Same Site Strict for cookies.
- Should I rebuild the express auth server to Next js?
- Is there anything I need to change in the deployment commands?

**Comment by amyegan** [comment]
Hi @aly8i! Are you able to share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) so I can get a better look at the source of the problem?

**Reply by aly8i** [reply]
Thank you for landing me a hand @amyegan click [here](https://github.com/aly8i/resturant) to view the source code of my project.

**Reply by amyegan** [reply]
I'm getting a little confused between your description of the problem and the code you shared. Sorry, I must be misunderstanding something about your situation. I can't really run the example code because of dependency conflicts. 😬 

It would be worth reading through ["Why does my Serverless Function work locally but not when deployed?"](https://vercel.com/guides/why-does-my-serverless-function-work-locally-but-not-when-deployed) to get a list of troubleshooting steps for Serverless Function errors. You may need to look at [enabling CORS](https://vercel.com/guides/how-to-enable-cors) in your app and/or in your Heroku express auth server. Or double check that your environment variables are set correctly for your deployment environment.

I find [function logs](https://vercel.com/docs/concepts/deployments/logs#function-logs) very helpful for debugging Serverless Function errors since they let me view output, even from plain `console.log({data})` kinds of statements, to see more details about what is succeeding and what is failing. Definitely recommend checking them out that if you have not already.

There is a guide available for [using Express with Vercel](https://vercel.com/guides/using-express-with-vercel) if you want to move the server, but there are some drawbacks to consider (which are explained in the linked guide). I can't guarantee that moving or rewriting the server would solve the problem though, since I haven't been able to see exactly what is causing it. ☹️

**Reply by aly8i** [reply]
@amyegan I think my problem is lying between these 4 possibilities. While sending some server requests, it is catching an error and breaking the application. Because on page load I am sending requests like signup, if it catches an error, I send login since the user exists, then continue normally if both fail. Another possibility is that these requests need some time and they are fired on getserversideprops so it is taking time to to generate some page info and because of that, the application is breaking. Because I've read something related to a limited allowed time for responses. plus I've put CI to false in the environment variables. I don't think the problem is with the express server tho. since in the logs while sending get request to api/statics to get some static data is failing too. is it maybe the environment variables? because the request endpoint is like so myapp.vercel.app/api/static should I keep it as localhost:3000/api/static for example and Vercel will do the route switching?

**Reply by amyegan** [reply]
The [execution timeout](https://vercel.com/docs/concepts/functions/serverless-functions#execution-timeout) error could be happening if the backend service takes a long time to return. The [timeout limit](https://vercel.com/docs/concepts/limits/overview#serverless-function-execution-timeout) varies by plan type:

> The maximum execution timeout is 10 seconds when deployed on a Personal Account (Hobby plan). For Teams, the execution timeout is 60 seconds (Pro plan) or 900 seconds (Enterprise plan).

I would definitely recommend double checking your environment variables to make sure access tokens aren't missing or expired. It's a fairly common cause of connection problems, which can look like simple timeout errors.

You shouldn't keep absolute `localhost:3000/api/static` URLs hardcoded. You can use the [`VERCEL_URL` environment variable](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) to get the appropriate URL for each environment.

**Reply by aly8i** [reply]
@amyegan In my case, I didn't hardcode the URLs, and accesstokens are valid and refreshed and even requests that authorize get request without accessTokens are failing. I've declared NEXT_PUBLIC_BASEURL = https://myapp.vercel.app and the request is like so axios.get(`${process.env.NEXT_PUBLIC_BASEURL}/api/static`); it works fine locally. Must I use VERCEL_URL literally and only while defining my URL?

**Reply by amyegan** [reply]
`NEXT_PUBLIC_VERCEL_URL` should work as well. If you can share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can run and deploy, that will help me get a better understanding of what is going wrong.

**Reply by aly8i** [reply]
@amyegan I found out that on my localhost while running "next build" then "next start", it is giving the same internal server error output as on my deployment. However, while running "next dev" the output is displayed correctly without errors on my localhost. So, I think my problem is that vercel is using the first procedure for deployment. So I need to configure vercel to use next dev instead of next start or configure the next start to behave as next dev. how can I do this?

**Reply by amyegan** [reply]
All deployments use the `build` step. You cannot deploy a **production** app with the `dev` command because of how the command runs for local development purposes.

If you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), I would be happy to take a closer look and help you debug.

**Comment by aly8i** [comment]
@amygen on production axios is giving this error related to "xsrf-token".

`AxiosError: connect ECONNREFUSED ::1:3000
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1237:16) {
  port: 3000,
  address: '::1',
  syscall: 'connect',
  code: 'ECONNREFUSED',
  errno: -4078,
  config: {
    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },
    adapter: [Function: httpAdapter],
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    env: { FormData: [Function] },
    validateStatus: [Function: validateStatus],
    headers: {
      Accept: 'application/json, text/plain, */*',
      'User-Agent': 'axios/0.27.2'
    },
    method: 'get',
    url: 'http://localhost:3000/api/products',
    data: undefined
  },
  request: <ref *1> Writable {
    _writableState: WritableState {
      objectMode: false,
      highWaterMark: 16384,
      finalCalled: false,
      needDrain: false,
      ending: false,
      ended: false,
      finished: false,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: true,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: null,
      writelen: 0,
      afterWriteTickInfo: null,
      buffered: [],
      bufferedIndex: 0,
      allBuffers: true,
      allNoop: true,
      pendingcb: 0,
      constructed: true,
      prefinished: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: true,
      errored: null,
      closed: false,
      closeEmitted: false,
      [Symbol(kOnFinished)]: []
    },
    _events: [Object: null prototype] {
      response: [Function: handleResponse],
      error: [Function: handleRequestError],
      socket: [Function: handleRequestSocket]
    },
    _eventsCount: 3,
    _maxListeners: undefined,
    _options: {
      maxRedirects: 21,
      maxBodyLength: 10485760,
      protocol: 'http:',
      path: '/api/products',
      method: 'GET',
      headers: [Object],
      agent: undefined,
      agents: [Object],
      auth: undefined,
      hostname: 'localhost',
      port: '3000',
      nativeProtocols: [Object],
      pathname: '/api/products'
    },
    _ended: true,
    _ending: true,
    _redirectCount: 0,
    _redirects: [],
    _requestBodyLength: 0,
    _requestBodyBuffers: [],
    _onNativeResponse: [Function (anonymous)],
    _currentRequest: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      maxRequestsOnConnectionReached: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      _closed: false,
      socket: [Socket],
      _header: 'GET /api/products HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'User-Agent: axios/0.27.2\r\n' +
        'Host: localhost:3000\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: nop],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/api/products',
      _ended: false,
      res: null,
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'localhost',
      protocol: 'http:',
      _redirectable: [Circular *1],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype],
      [Symbol(kUniqueHeaders)]: null
    },
    _currentUrl: 'http://localhost:3000/api/products',
    [Symbol(kCapture)]: false
  }
}`

**Comment by aly8i** [comment]
@amyegan  this is the only token/cookie I'm using :
      `res.setHeader("Set-Cookie",cookie.serialize("accessToken", access, {
        maxAge: 2*24*60*60*1000,
        sameSite: "strict",
        httpOnly: true,
        path: "/",
      })
    );`

**Comment by aly8i** [comment]
no such error in `npm run dev` these are generated while doing `npm run build` then `npm run start`

**Comment by aly8i** [comment]
@amyegan Do warnings break the production? here are some warnings while building my application:
`65:8  Warning: React Hook useEffect has missing dependencies: 'dispatch' and 'options'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps`
`20:5  Warning: img elements must have an alt prop, either with meaningful text, or an empty string for decorative images.  jsx-a11y/alt-text `
`20:5  Warning: Do not use `<img>` element. Use `<Image />` from `next/image` instead. See: https://nextjs.org/docs/messages/no-img-element  @next/`
`35:5  Warning: React Hook useEffect has missing dependencies: 'server.interceptors.request' and 'token'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps`
`18:7  Warning: React Hook useEffect has a missing dependency: 'deliverys'. Either include it or remove the dependency array. If 'setDeliveryList' needs the current value of 'deliverys', you can also switch to useReducer instead of useState and read 'deliverys' in the reducer.  react-hooks/exhaustive-deps`
